### ISSUE TITLE
HTTP parse error, malformed request (127.0.0.1): #<Puma::HttpParserError: Invalid HTTP format, parsing fails.>

### ISSUE THREAD

**Issue by anon987654321** [puma/puma#2232]
Hello,

I'm on OpenBSD with a single IP. I'm using OpenBSD's [httpd](https://man.openbsd.org/httpd.conf.5) with FastCGI listening on `puma.sock`. After running:
```
# puma --bind tcp://127.0.0.1 --control-url unix:///var/www/run/puma.sock --config-file /var/www/htdocs/mypumasite/config/puma.rb --environment production
```

I get this when attempting to visit http://mypumasite.com/ from an incognito window, `config.force_ssl` disabled:
```
Puma starting in single mode...
* Version 4.3.3 (ruby 2.6.6-p146), codename: Mysterious Traveller
* Min threads: 5, max threads: 5
* Environment: production
* Listening on tcp://127.0.0.1:35727
* Starting control server on unix:///var/www/run/puma.sock
Use Ctrl-C to stop
2020-04-24 09:26:19 +0200: HTTP parse error, malformed request (127.0.0.1): #<Puma::HttpParserError: Invalid HTTP format, parsing fails.>
---
```

I do host an SSL-enabled static site as well. For that I used this tutorial: [Enable HTTPS with acme-client(1) and Let’s Encrypt on OpenBSD](https://romanzolotarev.com/openbsd/acme-client.html).

Here's my `/etc/httpd.conf`:
```
server "mypumasite.com" {
  listen on * port 80
  alias "www.mypumasite.com" 
  root "/htdocs/mypumasite"
  fastcgi socket "/run/mypumasite.sock"
}

server "mystaticsite.com" {
  listen on * port 80
  listen on * tls port 443
  alias "www.mystaticsite.com"
  root "/htdocs/mystaticsite"
  location "/.well-known/acme-challenge/*" {
    root "/acme"
    request strip 2
  }
  tls {
    certificate "/etc/ssl/mystaticsite.pem"
    key "/etc/ssl/private/mystaticsite.key"
  }
}
```

The DNS server for `mypumasite.com` is `ns.mystaticsite.com`. It has a [CAA record](https://letsencrypt.org/docs/caa/) but only for `mystaticsite.com`.

That's about as verbose as I can get. Hope you guys can help.

Thanks.

**Comment by dentarg** [comment]
> I get this when attempting to visit http://mypumasite.com/ from an incognito window

Using Google Chrome I assume? Does it always happen? Can you reproduce the error with another user agent, e.g. curl?

**Comment by anon987654321** [comment]
@dentarg Indeed, unfortunately the error remains the same with curl as well.

**Comment by nateberkopec** [comment]
Can you post any of `/var/www/htdocs/mypumasite/config/puma.rb`?

**Comment by anon987654321** [comment]
Sure! I haven't touched it though, it's the default from Rails 6.0.2.2:
```
# Puma can serve each request in a thread from an internal thread pool.
# The `threads` method setting takes two numbers: a minimum and maximum.
# Any libraries that use thread pools should be configured to match
# the maximum value specified for Puma. Default is set to 5 threads for minimum
# and maximum; this matches the default thread size of Active Record.
#
max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
threads min_threads_count, max_threads_count

# Specifies the `port` that Puma will listen on to receive requests; default is           3000.
#
port        ENV.fetch("PORT") { 3000 }

# Specifies the `environment` that Puma will run in.
#
environment ENV.fetch("RAILS_ENV") { "development" }

# Specifies the `pidfile` that Puma will use.
pidfile ENV.fetch("PIDFILE") { "tmp/pids/server.pid" }

# Specifies the number of `workers` to boot in clustered mode.
# Workers are forked web server processes. If using threads and workers together
# the concurrency of the application would be max `threads` * `workers`.
# Workers do not work on JRuby or Windows (both of which do not support
# processes).
#
# workers ENV.fetch("WEB_CONCURRENCY") { 2 }

# Use the `preload_app!` method when specifying a `workers` number.
# This directive tells Puma to first boot the application and load code
# before forking the application. This takes advantage of Copy On Write
# process behavior so workers use less memory.
#
# preload_app!

# Allow puma to be restarted by `rails restart` command.
plugin :tmp_restart
```

**Comment by nateberkopec** [comment]
The debug output here is pretty unhelpful. I opened a new issue at #2235.

**Comment by nateberkopec** [comment]
In short, I think this is a problem w/your setup, not Puma, but we could really improve our debug logging to help you figure out exactly what's going on.

**Comment by dentarg** [comment]
> After running:
> 
> ```
> # puma --bind tcp://127.0.0.1 --control-url unix:///var/www/run/puma.sock --config-file /var/www/htdocs/mypumasite/config/puma.rb --environment production
> ```

Is this the correct command? `puma` doesn't have the `--config-file` flag (`pumactl` does, but `pumactl` does not have the `--bind` flag)

I'm trying to reproduce this (in OpenBSD). Are you using the OpenBSD 6.6 release?

**Comment by dentarg** [comment]
Okay, see https://man.openbsd.org/httpd.conf.5#socket

> httpd passes HTTP requests to a FastCGI handler listening on the socket _socket_. 

I suspect the issue is that Puma isn't a FastCGI handler. You need a web server that can act as reverse proxy (e.g. nginx) I don't think httpd can do that.

**Comment by anon987654321** [comment]
Oh I see! I think [relayd](https://bsd.plumbing/) would be a good match for that, it's made by OpenBSD as well.

Thank you so much for helping me get to the bottom of this!

**Comment by anon987654321** [comment]
Feel free to chip in if you too want to see a working Puma configuration for OpenBSD:

[Reddit r/openbsd: relayd/httpd/fastcgi reverse proxy fail](https://www.reddit.com/r/openbsd/comments/g8cyi4/relaydhttpdfastcgi_reverse_proxy_fail/)

**Comment by anon987654321** [comment]
Hey! For what it's worth, I finally managed to get it working. Feedback welcome:

https://gist.github.com/anon987654321/4532cf8d6c59c1f43ec8973faa031103

**Comment by sekroots** [comment]
Hello. This is not an OpenBSD issue. I have an ruby on rails app runing under docker and I am facing this issue. How do I solve it?

`WARN[0000] network default: network.external.name is deprecated. Please set network.name with external: true 
web_1  | delayed_job: process with pid 14 started.
web_1  | Puma starting in single mode...
web_1  | * Version 3.12.6 (ruby 3.0.6-p216), codename: Llamas in Pajamas
web_1  | * Min threads: 5, max threads: 5
web_1  | * Environment: production
web_1  | * Listening on tcp://0.0.0.0:3003
web_1  | Use Ctrl-C to stop
web_1  | 2024-03-04 08:44:24 +0000: HTTP parse error, malformed request (): #<Puma::HttpParserError: Invalid HTTP format, parsing fails.>
web_1  | ---
`

**Comment by MSP-Greg** [comment]
JFYI, Puma 3.12.6 currently accounts for less than 2% of Puma downloads.  It was also released before the release of Ruby 3.0.0.

Personally, I haven't worked with Puma 3.1 for quite a while, and I don't recall the transition from Ruby 2.7 to Ruby 3.0.  Maybe others can help.

**Comment by anon987654321** [comment]
That's right, on OpenBSD relayd is best suited for this. Check out my
setup, where httpd only aids in generating the TLS certificates. Relayd
does the reverse proxying and TLS termination before handing it all over to
Puma/whatever.

https://gist.github.com/basicfeatures/7d1448aad6fadc0def4ffee443f26dab

I'm about to release a version 2.0 fully battle-tested version of this, so
if it seems interesting, let me know and I'll share the beta.

Check out my Docker article while at it:

***@***.***/openbsd-vs-docker-and-linux-deploying-ruby-on-rails-in-production-320c90bcb934

I hope this all finds y'all well.

--BF
